src/librustc_typeck/check/wfcheck.rs

```diff
@@ -354,19 +343,112 @@ impl<'a, 'gcx> CheckTypeWellFormedVisitor<'a, 'gcx> {
+                                def.has_default && def.index >= generics.parent_count() as u32;
+        let defaulted_params = generics.types.iter().cloned().filter(&is_our_default);
+        // Check that defaults are well-formed. See test `type-check-defaults.rs`.
+        // For example this forbids the declaration:
+        // struct Foo<T = Vec<[u32]>> { .. }
+        // Here the default `Vec<[u32]>` is not WF because `[u32]: Sized` does not hold.
+        for d in defaulted_params.map(|p| p.def_id) {
+            fcx.register_wf_obligation(fcx.tcx.type_of(d), fcx.tcx.def_span(d), self.code.clone());
+        }
+
+        // Check that trait predicates are WF when params are substituted by their defaults.
+        // We don't want to overly constrain the predicates that may be written but we
+        // want to catch obviously wrong cases such as `struct Foo<T: Copy = String>`
+        // or cases where defaults don't work together such as:
+        // `struct Foo<T = i32, U = u8> where T: into<U>`
+        // Therefore we check if a predicate in which all type params are defaulted
+        // is WF with those defaults simultaneously substituted.
+        // For more examples see tests `defaults-well-formedness.rs` and `type-check-defaults.rs`.
+        //
+        // First we build the defaulted substitution.
+        let mut defaulted_params = Vec::new();
+        let substs = ty::subst::Substs::for_item(fcx.tcx, def_id, |def, _| {
+                // All regions are identity.
+                fcx.tcx.mk_region(ty::ReEarlyBound(def.to_early_bound_region_data()))
+            }, |def, _| {
+                if !is_our_default(def) {
+                    // Identity substitution.
+                    fcx.tcx.mk_param_from_def(def)
+                } else  {
+                    // Substitute with default.
+                    defaulted_params.push(def.index);
+                    fcx.tcx.type_of(def.def_id)
+                }
+            });
+        let defaulted_params = &defaulted_params;
+        // Now we build the substituted predicates.
+        for &pred in predicates.predicates.iter() {
+            struct HasNonDefaulted<'a> { defaulted_params: &'a Vec<u32> }
+            impl<'tcx, 'a> ty::fold::TypeVisitor<'tcx> for HasNonDefaulted<'a> {
+                fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+                    match t.sty {
+                        ty::TyParam(p) => !self.defaulted_params.contains(&p.idx),
+                        _ => t.super_visit_with(self)
+                    }
+                }
+            }
+            // If there is a non-defaulted param in the predicate, don't check it.
+            if pred.visit_with(&mut HasNonDefaulted { defaulted_params }) {
+                continue;
+            }
+            let substituted_pred = pred.subst(fcx.tcx, substs);
+            // In trait defs, don't check `Self: Sized` when `Self` is the default.
```

Comment:
Self never has a default, right? So do we have to explicit skip `Self: Sized`?

Reply:
`Self` never has a default but it can appear as a default. In general type parameters may sneak into these predicates through dependent defaults, for example:
```rust
struct Dependent<T: Copy, U = T>(T, U) where U: Copy; // Valid.
struct Dependent<T, U = T>(T, U) where U: Copy; // Not valid.
```